dev/PowerNotifications/PowerNotifications.cpp

```diff
@@ -0,0 +1,501 @@
+            case PowerSourceStatusFn:      check_hresult(RegisterPowerConditionChangedListener(PowerSourceStatusChanged_Callback, &m_powerConditionHandle));       break;
+            case DisplayStatusFn:          check_hresult(RegisterDisplayStatusChangedListener(DisplayStatusChanged_Callback, &m_displayStatusHandle));             break;
+            case SystemIdleStatusFn:       check_hresult(RegisterSystemIdleStatusChangedListener(SystemIdleStatusChanged_Callback, &m_systemIdleStatusHandle));    break;
+            case PowerSchemePersonalityFn: check_hresult(RegisterPowerSchemePersonalityChangedListener(PowerSchemePersonalityChanged_Callback, &m_powerSchemePersonalityHandle)); break;
+            case UserPresenceStatusFn:     check_hresult(RegisterUserPresenceStatusChangedListener(UserPresenceStatusChanged_Callback, &m_userPresenceStatusHandle));             break;
+            case SystemAwayModeStatusFn:   check_hresult(RegisterSystemAwayModeStatusChangedListener(SystemAwayModeStatusChanged_Callback, &m_systemAwayModeStatusHandle));       break;
+            default: throw E_NOTIMPL;
+        }
+    }
+
+    void PowerManager::UnregisterListener(PowerFunction const& fn)
+    {
+        switch (fn) {        
+            case BatteryStatusFn:          
+            case PowerSupplyStatusFn:      
+            case RemainingChargePercentFn: check_hresult(UnregisterCompositeBatteryStatusChangedListener(m_batteryStatusHandle)); break;
+            case EnergySaverStatusFn:      check_hresult(UnregisterEnergySaverStatusChangedListener(m_energySaverStatusHandle));  break;
+            case RemainingDischargeTimeFn: check_hresult(UnregisterDischargeTimeChangedListener(m_dischargeTimeHandle));          break;
+            case PowerSourceStatusFn:      check_hresult(UnregisterPowerConditionChangedListener(m_powerConditionHandle));        break;
+            case DisplayStatusFn:          check_hresult(UnregisterDisplayStatusChangedListener(m_displayStatusHandle));          break;
+            case SystemIdleStatusFn:       check_hresult(UnregisterSystemIdleStatusChangedListener(m_systemIdleStatusHandle));    break;
+            case PowerSchemePersonalityFn: check_hresult(UnregisterPowerSchemePersonalityChangedListener(m_powerSchemePersonalityHandle)); break;
+            case UserPresenceStatusFn:     check_hresult(UnregisterUserPresenceStatusChangedListener(m_userPresenceStatusHandle));         break;
+            case SystemAwayModeStatusFn:   check_hresult(UnregisterSystemAwayModeStatusChangedListener(m_systemAwayModeStatusHandle));     break;
+            default: throw E_NOTIMPL;
+        }
+    }
+
+    void PowerManager::RemoveCallback(PowerFunction const& fn, EventToken const& token)
+    {
+        auto eventObj = GetEventObj(fn);
+        eventObj.get().remove(token);
+        std::scoped_lock<std::mutex> lock(m_mutex);
+        //Check prior registrations
+        if (NotAlreadyRegisteredForEvents(eventObj))
```

Comment:
This condition seems backwards? If the event has already been registered, then we should unregister.